### PR TITLE
Fix:Kubectl reports wrong status for evicted pod when container statuses are synced

### LINKED ISSUE #105467: Kubectl reports wrong status for evicted pod when container statuses are synced after
#### What happened:

In Kubernetes <1.22, Evicted pods would not have any container status synced. Kubectl displayed them as "Evicted".

In 1.22+, we fixed a bug where we now sync out the last known container status from the runtime on Evicted pods. However, that status syncs after we update the pod status. As soon as the runtime status syncs, since it is more recent, it takes precedence in the display, and the pod's status updates from Evicted to whatever the container's status is (Error, ContainerStatusUnknown, etc.)

#### What you expected to happen:

So long as the pod status stays Evicted with phase Failed, I expected kubectl to print Evicted.

#### How to reproduce it (as minimally and precisely as possible):

See https://github.com/kubernetes/kubernetes/issues/105358#issue-1011532980 and https://github.com/kubernetes/kubernetes/issues/105358#issuecomment-932470027

#### Anything else we need to know?:

/sig cli node

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by 249043822**
In Kubernetes <1.22, Evicted pods would not have any container status synced. Kubectl displayed them as "Evicted".

In 1.22+, we fixed a bug where we now sync out the last known container status from the runtime on Evicted pods. However, that status syncs after we update the pod status. As soon as the runtime status syncs, since it is more recent, it takes precedence in the display, and the pod's status updates from Evicted to whatever the container's status is (Error, ContainerStatusUnknown, etc.)

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

Now we sync out the last known container status from the runtime even the pod has been evicted, preempted or teminated by node shutdown, we should fix kubectl display the correct pod Status

#### Which issue(s) this PR fixes:

Fixes #105467

#### Special notes for your reviewer:
/cc @ehashman @smarterclayton 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/printers/internalversion/printers.go [hunk 1]
```diff
@@ -832,7 +832,12 @@ func printPod(pod *api.Pod, options printers.GenerateOptions) ([]metav1.TableRow
 			}
 		}
 	}
-
+	// When pod was evicted, preempted or even terminated in response to imminent node shutdown,
```
Comment (smarterclayton) [inline pkg/printers/internalversion/printers.go:835]: This is probably the minimal change, but a part of me wonders whether this is about specific reasons, or about trusting the kubelet to give us an accurate reason.  The corresponding changes on the kubelet when this regressed were to make the Reason more authoritative (but it is not always so).  I wonder if we should:

1. make the reason calculated for container go to a variable called containerReason
2. have a switch statement here based on pod status reason, taking this logic and some of the existing logic that either defaults or overrides container reason
3. describe formally in pod status api reason well known reasons that supercede container reason

The problem is container reason and pod reason aren't really the same thing - the reason we return from the api is supposed to be a summary, while the container reason logic here in the printer is trying to present some useful info that would help the user in advance of the reason.

Generally, I think container reason should never be set IF pod set a reason, but we'd need to check that pod reason is that accurate in the kubelet, and then clarify in api docs.  I think that change is a bit more dramatic than this, but would help future regressions and would simplify the code (pod reason >> container reason unless unset)

Comment (249043822) [reply pkg/printers/internalversion/printers.go:835]: @smarterclayton Generally, a pod is set status.Reason as it becomes terminal(eg. Evicted, Rejected by admit...), And container is set state.Reason only if it is not in running state(terminal or waiting). Current, the logic in the printer is that using containerstatus.Reason to override podstatus.Reason first, but if we now use pod reason to override container reason,  this is completely reversed.  
Still, I think that keeping the original reason of containerstatus can give more details for users, but what you said still make sense to me,  so shall we only change the describe logic in printer, not change the kubelet convert api?

Comment (249043822) [reply pkg/printers/internalversion/printers.go:835]: > This is probably the minimal change, but a part of me wonders whether this is about specific reasons, or about trusting the kubelet to give us an accurate reason. The corresponding changes on the kubelet when this regressed were to make the Reason more authoritative (but it is not always so). I wonder if we should:
> 
> 1. make the reason calculated for container go to a variable called containerReason
> 2. have a switch statement here based on pod status reason, taking this logic and some of the existing logic that either defaults or overrides container reason
> 3. describe formally in pod status api reason well known reasons that supercede container reason
> 

we may do this in the func below https://github.com/kubernetes/kubernetes/blob/bdb9c0824c5c08d6dc7128a27e384011cfd19e64/pkg/kubelet/kubelet_pods.go#L1627


> The problem is container reason and pod reason aren't really the same thing - the reason we return from the api is supposed to be a summary, while the container reason logic here in the printer is trying to present some useful info that would help the user in advance of the reason.
> 
> Generally, I think container reason should never be set IF pod set a reason, but we'd need to check that pod reason is that accurate in the kubelet, and then clarify in api docs. I think that change is a bit more dramatic than this, but would help future regressions and would simplify the code (pod reason >> container reason unless unset)

**Review by yangjunmyfm192085 — COMMENTED** [review]
/cc @[249043822](https://github.com/249043822)

File: pkg/kubelet/status/status_manager.go [hunk 2]
```diff
@@ -888,3 +891,32 @@ func NeedToReconcilePodReadiness(pod *v1.Pod) bool {
...
+		// update init container reason
+		for i := range status.InitContainerStatuses {
+			cstatus := &status.InitContainerStatuses[i]
+			if cstatus.State.Terminated != nil {
+				cstatus.State.Terminated.Reason = status.Reason
+			} else if cstatus.State.Waiting != nil {
+				cstatus.State.Waiting.Reason = status.Reason
+			}
+
+		}
```
Comment (Dragoncell) [inline pkg/kubelet/status/status_manager.go:919]: This part of code is kind of duplicate as above one for update the container reason, maybe could create a separate function and pass the container/initContainer status as variable to function to achieve the update.

**Review by marquiz — COMMENTED** [review]
apart from @Dragoncell's comment, looks good to me

**Comment by Monokaix** [comment]
@smarterclayton can we resolve this?

File: pkg/kubelet/status/status_manager.go [hunk 3]
```diff
@@ -496,7 +496,11 @@ func (m *manager) updateStatusInternal(pod *v1.Pod, status v1.PodStatus, forceUp
 	}
 
 	normalizeStatus(pod, &status)
-
+	// Since kubectl prefers using container reason to supercede pod reason to display pod STATUS,
+	// so when pod was evicted, preempted or even terminated in response to imminent node shutdown,
+	// and pod phase is Failed, then we override container reason by pod reason unless unset.
+	// See issue #105467 for detail.
+	normalizeContainerResaon(pod, &status)
```
Comment (swatisehgal) [inline pkg/kubelet/status/status_manager.go:503]: NIT:
```suggestion
	normalizeContainerReason(pod, &status)
```

File: pkg/kubelet/status/status_manager.go [hunk 4]
```diff
@@ -888,3 +891,32 @@ func NeedToReconcilePodReadiness(pod *v1.Pod) bool {
 	}
 	return false
 }
+
+func normalizeContainerResaon(pod *v1.Pod, status *v1.PodStatus) *v1.PodStatus {
```
Comment (swatisehgal) [inline pkg/kubelet/status/status_manager.go:895]: NIT:
```suggestion
func normalizeContainerReason(pod *v1.Pod, status *v1.PodStatus) *v1.PodStatus {
```

File: pkg/kubelet/status/status_manager.go (hunk 2 above)
Comment (swatisehgal) [inline pkg/kubelet/status/status_manager.go:919]: I agree with @Dragoncell. It would be better to add a helper function here.

**Review by swatisehgal — COMMENTED** [review]
@249043822 Are you still working on this? It would be great to move this forward.

Can you please address the review comments and rebase the PR. Thanks!

File: pkg/kubelet/status/status_manager.go [hunk 5]
```diff
@@ -888,3 +891,32 @@ func NeedToReconcilePodReadiness(pod *v1.Pod) bool {
...
+	// When pod was evicted, preempted, admitted failure or even terminated in response to imminent node shutdown,
+	// the pod became Failed, kubelet would set accurate reason for the pod.
+	// If the restart policy is Never and the pod becomes failed, kubelet would not set reason, that is not the case.
+	if status.Phase == v1.PodFailed && len(status.Reason) != 0 {
+		// update container reason
```
Comment (kannon92) [inline pkg/kubelet/status/status_manager.go:900]: Do we need to account for ephemeral containers also?

**Comment by hshiina** [comment]
If this issue mentions only `STATUS` displayed by `kubectl get pods`, it might be better to fix here:
https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/pkg/printers/internalversion/printers.go#L963-L966
It looks a little strange for me that this fix overwrites containers with a pod while the `printPod()` overwrites a pod with containers.

**Pull request closed without merging**